text/0000-Declarative-Services-GetBundleContext.md

```diff
@@ -0,0 +1,49 @@
+- Start Date: 2023-09-18
+- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: https://github.com/CppMicroServices/CppMicroServices/issues/491
+
+# Supporting GetBundleContext inside Declarative Services
+
+## Summary
+
+Currently, the freestanding `GetBundleContext` function generally does not work when used inside a declarative service. This is for two reasons: the necessary data is usually not present in the bundle object, and even when present, that data is often not initialized. This RFC is for the resolution of these two deficiencies.
+
+## Motivation
+
+As Declarative Services is a layer on top of the Core Framework, developers most likely would not expect features of the CF like `GetBundleContext` to stop working when using DS. This issue also does not always occur, and is not noted in the documentation. From a bundle developer's perspective, this is a bug in CppMicroServices.
+
+## Detailed design
+
+Declarative Services offers two benefits that are relevant to this issue: it generates boilerplate code for the user, and it allows lazy loading to optimize memory usage. Unfortunately, some things were missed when implementing these features. Specifically:
+
+1. Use of the macro that provides data storage for `GetBundleContext` is discouraged when using Declarative Services, but its functionality has not been replaced.
+2. A logic error in the immediate bundle loader means that this data storage is not initialized unless `"bundle.activator"` is set to `true` in the bundle manifest.
+3. The lazy bundle loader, which is the default loader, does not initialize this data storage under any circumstances.
+
+The data storage and its accessor functions that allow `GetBundleContext` to work are provided when the user adds the `CPPMICROSERVICES_INITIALIZE_BUNDLE` macro to their bundle code (`BundleInitialization.h`). While this is required without DS, our current documentation shows it not being used with DS. Because the DS Code Generation Tool already provides a facility for inserting code into the bundle, the tool will be modified to insert the necessary code, effectively providing the contents of the `CPPMICROSERVICES_INITIALIZE_BUNDLE` macro for the developer automatically.
```

Comment:
Same issue here. Here's the definition of the macro:

    #    define CPPMICROSERVICES_INITIALIZE_BUNDLE                                                                       \
        std::atomic<cppmicroservices::BundleContextPrivate*> US_CTX_INS(US_BUNDLE_NAME) {};                          \
                                                                                                                     \
        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)();                        \
        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)()                         \
        {                                                                                                            \
            return US_CTX_INS(US_BUNDLE_NAME).load();                                                                \
        }                                                                                                            \
                                                                                                                     \
        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx); \
        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx)  \
        {                                                                                                            \
            US_CTX_INS(US_BUNDLE_NAME).store(ctx);                                                                   \
        }

Where is data storage allocated? I see he declaration of an atomic BundleContextPrivate*, and a function to get its value and set its value. No allocations, no storage provided.

Reply:
I got tripped up by the same thing when I first read this code. The first line declares a brace-initialized global variable - the US_CTX_INS(US_BUNDLE_NAME) macros expand to a variable name. Though the parens and braces make it look like an empty function at first glance. Note the extra pair of parens for the function declarations.

Reply:
Yep. I got that. The first line declares a global variable that is just a pointer to an object. Nothing is allocated. The pointer (which points to the allocated memory) must be set with a call to the US_SET_CTX_FUNC().

Reply:
I wrote this thinking about the std::atomic object, whose purpose is storing some data we want later (8 bytes of pointer). Though conceptually we care more about the BundleContextPrivate it points to. I'll adjust the wording to focus on that instead.

Reply:
Perfect. :-)